common/defs.go

```diff
@@ -197,58 +199,74 @@ func (p *Policy) Validate() error {
+			}
+		} else {
+			if len(r.Ports) > 0 || len(r.PortRanges) > 0 {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: ICMP protocol is specified but ports are also specified.", ruleNo))
+			}
+			if r.IcmpType < 0 || r.IcmpType > MaxIcmpType {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP type: %d.", ruleNo, r.IcmpType))
+			}
+			switch r.IcmpType {
+			case 3: // Destination unreachable
+				if r.IcmpCode < 0 || r.IcmpCode > 15 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
-				if r.IcmpType < 0 || r.IcmpType > MaxIcmpType {
-					errMsg += fmt.Sprintf("Invalid ICMP type: %d. ", r.IcmpType)
+			case 4: // Source quench
+				if r.IcmpCode != 0 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
-				switch r.IcmpType {
-				case 3: // Destination unreachable
-					if r.IcmpCode < 0 || r.IcmpCode > 15 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 4: // Source quench
-					if r.IcmpCode != 0 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 5: // Redirect
-					if r.IcmpCode < 0 || r.IcmpCode > 3 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 11: // Time exceeded
-					if r.IcmpCode < 0 || r.IcmpCode > 1 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				default:
-					if r.IcmpCode != 0 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-
+			case 5: // Redirect
+				if r.IcmpCode < 0 || r.IcmpCode > 3 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
+				}
+			case 11: // Time exceeded
+				if r.IcmpCode < 0 || r.IcmpCode > 1 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
+			default:
+				if r.IcmpCode != 0 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
+				}
+
 			}
+
+		}
+	}
+	for i, endpoint := range p.AppliedTo {
+		epNo := i + 1
+		if endpoint.TenantExternalID == "" && endpoint.TenantID == 0 {
+			errMsg = append(errMsg, fmt.Sprintf("In applied_to entry at %d, at least one of tenant_id or tenant_external_id must be specified.", epNo))
 		}
 	}
-	if errMsg == "" {
+	for i, endpoint := range p.Peers {
+		epNo := i + 1
+		if endpoint.SegmentID != 0 || endpoint.SegmentExternalID != "" {
+			if endpoint.TenantExternalID == "" && endpoint.TenantID == 0 {
+				errMsg = append(errMsg, fmt.Sprintf("In peers entry at %d, since segment_external_id is specified, at least one of tenant_id or tenant_external_id must be specified.", epNo))
```

Comment:
I'm not feeling so good about us schlepping the external ID around in our internal data structures. What is it good for? Who needs it? Would it not be cleaner to maintain a lookup table via the adaptor code somewhere, so that we can keep our code clear of external IDs?